File: ThreeEngine/lights/DirectionalLight.cpp
#include <GL/glew.h>
#include "DirectionalLight.h"
#include "DirectionalLightShadow.h"

DirectionalLight::DirectionalLight()
	: m_constant(sizeof(ConstDirectionalLight), GL_UNIFORM_BUFFER)
{

}

DirectionalLight::~DirectionalLight()
{

}

void DirectionalLight::setShadow(bool enable, int map_width, int map_height)
{
	if (enable)
	{
		shadow = std::unique_ptr<DirectionalLightShadow>(new DirectionalLightShadow(this));
		shadow->update_shadowmap(map_width, map_height);
	}
	else
	{
		shadow = nullptr;
	}
}

void DirectionalLight::setShadowProjection(float left, float right, float bottom, float top, float zNear, float zFar)
{
	if (shadow != nullptr)
	{
		shadow->setProjection(left, right, bottom, top, zNear, zFar);
	}
}

void DirectionalLight::SetShadowRadius(float radius)
{
	if (shadow != nullptr)
	{
		shadow->m_light_radius = radius;
	}
}

glm::vec3 DirectionalLight::direction()
{
	this->updateWorldMatrix(true, false);
	glm::vec3 pos_target = { 0.0f, 0.0f, 0.0f };
	if (target != nullptr)
	{
		pos_target = target->matrixWorld[3];
	}
	glm::vec3 position = matrixWorld[3];	
	return glm::normalize(position - pos_target);
}

void DirectionalLight::makeConst(ConstDirectionalLight& const_light)
{
	const_light.color = glm::vec4(color * intensity, 1.0f);
	const_light.origin = glm::vec4(position, 1.0f);
	const_light.direction = glm::vec4(direction(), 0.0f);
	const_light.has_shadow = shadow != nullptr ? 1 : 0;
	const_light.diffuseThresh = diffuse_thresh;
	const_light.diffuseHigh = diffuse_high;
	const_light.diffuseLow = diffuse_low;
	const_light.specularThresh = specular_thresh;
	const_light.specularHigh = specular_high;
	const_light.specularLow = specular_low;
}

void DirectionalLight::updateConstant()
{
	ConstDirectionalLight c;
	makeConst(c);
	m_constant.upload(&c);
}

void DirectionalLight::lookAtTarget()
{
	glm::vec3 abs_dir = glm::abs(direction());
	if (abs_dir.y < abs_dir.x)
	{
		if (abs_dir.z < abs_dir.y)
		{
			this->up = { 0.0f, 0.0f, 1.0f };
		}
		else
		{
			this->up = { 0.0f, 1.0f, 0.0f };
		}
	}
	else if (abs_dir.z < abs_dir.x)
	{
		this->up = { 0.0f, 0.0f, 1.0f };
	}
	else
	{
		this->up = { 1.0f, 0.0f, 0.0f };
	}

	glm::vec3 pos_target = { 0.0f, 0.0f, 0.0f };
	if (target != nullptr)
	{
		pos_target = target->matrixWorld[3];
	}
	lookAt(pos_target);
}

File: webgpu/client/engine/lights/DirectionalLight.js
import { Light } from "./Light.js"
import { Vector3 } from "../math/Vector3.js";

export class DirectionalLight extends Light
{
    constructor()
    {
        super();
        this.target = null;
        this.shadow = null;
        this.constant = engine_ctx.createBuffer0(80, GPUBufferUsage.UNIFORM|GPUBufferUsage.COPY_DST);
    }

    direction()
    {     
        this.updateWorldMatrix(true, false);
        let pos_target = new Vector3(0,0,0);
        if (this.target!=null)
        {
            pos_target.setFromMatrixPosition(this.target.matrixWorld);
        }
        let dir = new Vector3();
        dir.setFromMatrixPosition(this.matrixWorld);
        dir.sub(pos_target);
        dir.normalize();
        return dir;
    }

    lookAtTarget()
    {
        let dir = this.direction();
        if (Math.abs(dir.y)<Math.abs(dir.x))
        {
            if (Math.abs(dir.z)<Math.abs(dir.y))
            {
                this.up.set(0,0,1);
            }
            else
            {
                this.up.set(0,1,0);
            }
        }
        else if (Math.abs(dir.z)<Math.abs(dir.x))
        {
            this.up.set(0,0,1);
        }
        else
        {
            this.up.set(1,0,0);
        }    
        
        let pos_target = new Vector3(0,0,0);
        if (this.target!=null)
        {
            pos_target.setFromMatrixPosition(this.target.matrixWorld);
        }
        this.lookAt(pos_target);
    }

    updateConstant()
    {
        let uniform = new Float32Array(20);        

        let direction = this.direction();
        let position = new Vector3();
        position.setFromMatrixPosition(this.matrixWorld);        

        uniform[0] = this.color.r * this.intensity;
        uniform[1] = this.color.g * this.intensity;
        uniform[2] = this.color.b * this.intensity;
        uniform[3] = 1.0;
        uniform[4] = position.x;
        uniform[5] = position.y;
        uniform[6] = position.z;
        uniform[7] = 1.0;
        uniform[8] = direction.x;
        uniform[9] = direction.y;
        uniform[10] = direction.z;
        uniform[11] = 0.0;        
        uniform[12] = this.diffuse_thresh;
        uniform[13] = this.diffuse_high;
        uniform[14] = this.diffuse_low;
        uniform[15] = this.specular_thresh;
        uniform[16] = this.specular_high;
        uniform[17] = this.specular_low;
        engine_ctx.queue.writeBuffer(this.constant, 0, uniform.buffer, uniform.byteOffset, uniform.byteLength);
    }


}
